src/main/groovy/com/morpheusdata/upcloud/services/UpcloudApiService.groovy

```diff
@@ -0,0 +1,863 @@
+        def retryInterval = opts?.retryInterval ?: (1000l * 5l)
+        try {
+            def pending = true
+            def attempts = 0
+            while(pending) {
+                def serverDetail = getStorageDetails(authConfig, storageId)
+                if(serverDetail.success == true && serverDetail?.storage?.state == status) {
+                    rtn.success = true
+                    rtn.results = serverDetail.storage
+                    pending = false
+                    break
+                }
+                attempts ++
+                if(attempts > maxAttempts) {
+                    pending = false
+                } else {
+                    sleep(retryInterval)
+                }
+            }
+        } catch(e) {
+            log.error(e)
+        }
+        return rtn
+    }
+
+    static restoreSnapshot(Map authConfig, String stroageId) {
+        def rtn = [success:false]
+        try {
+            def callOpts = [:]
+            def callPath = "/storage/${stroageId}/restore"
+            def callResults = callApi(authConfig, callPath, callOpts, 'POST')
+            if(callResults.success == true) {
+                rtn.data = callResults.data
+                rtn.success = true
+            } else {
+                rtn.success = false
+            }
+        } catch (e) {
+            log.error "Error on restoreSnapshot: ${e}", e
+            rtn.success = false
+        }
+        return rtn
+    }
+
+    /*static listSnapshots(apiUrl, apiKey, opts){
+        def rtn = [success:false]
+        def headers = ['Content-Type':'application/json']
+        def results = callApi(apiUrl, '/v2/droplets/' + opts.server.externalId + '/snapshots', apiKey, [headers:headers, requestContentType:ContentType.JSON], Method.GET)
+        log.debug("got: ${results}")
+        rtn.success = results?.success
+        if(rtn.success == true) {
+            rtn.results = new groovy.json.JsonSlurper().parseText(results.content)
+            rtn.action = rtn.results?.action
+        }
+        return rtn
+    }*/
+
+    static callApi(Map authConfig, String path, Map opts = [:], String method) {
+        def apiUrl = authConfig.apiUrl
+        String username = authConfig.username
+        String password = authConfig.password
+        def apiVersion = authConfig.apiVersion ?: upcloudApiVersion
+        String apiPath = "${apiVersion}${path}"
+        String fullUrl = "${apiUrl}/${apiPath}"
+
+        String creds = "${username}:${password}".toString()
+        String authHeader = "Basic ${creds.getBytes().encodeBase64().toString()}".toString()
+
+        RequestOptions requestOptions = new RequestOptions()
+        if(opts.body) {
+           requestOptions.body = opts.body
+        }
+
+        if(opts.query) {
+            opts.query?.each { k, v ->
+                requestOptions.queryParams.(k) = v
+            }
+        }
+
+        requestOptions.headers = ['Authorization':authHeader, 'Content-Type':'application/json']
```

Comment:
For future proofing, process `opts.headers` the same as `opts.query`. This will also allow the overriding of the default headers set here.